Yank/restraints.py

```diff
@@ -96,330 +107,392 @@ def available_restraint_types():
     return available_restraints.keys()
 
 
-def create_restraints(restraint_type, topology, state, system, positions, receptor_atoms, ligand_atoms):
-    """
-    Initialize a receptor-ligand restraint class matching the specified restraint type name.
+def create_restraint(restraint_type, **kwargs):
+    """Factory of receptor-ligand restraint objects.
 
     Parameters
     ----------
     restraint_type : str
         Restraint type name matching a register (imported) subclass of ReceptorLigandRestraint.
-    topology : openmm.app.Topology
-        Reference topology for the complex
-    system : openmm.System
-        System containing all the forces and atoms that restraint will be added to
-    state : thermodynamics.ThermodynamicState
-        The thermodynamic state specifying temperature, pressure, etc. to which restraints are to be added
-    positions : simtk.unit.Quantity of natoms x 3 with units compatible with nanometers
-        Reference positions to use for imposing restraints
-    receptor_atoms : list of int
-        A complete list of receptor atoms
-    ligand_atoms : list of int
-        A complete list of ligand atoms
+    **kwargs
+        Parameters to pass to the restraint constructor.
 
     """
     available_restraints = available_restraint_classes()
-    if not (restraint_type in available_restraints):
-        raise ValueError("Restraint type '%s' unknown. Options are: %s" % (restraint_type, str(available_restraints.keys())))
+    if restraint_type not in available_restraints:
+        raise ValueError("Restraint type {} unknown. Options are: {}".format(
+            restraint_type, str(available_restraints.keys())))
     cls = available_restraints[restraint_type]
-    return cls(topology, state, system, positions, receptor_atoms, ligand_atoms)
+    return cls(**kwargs)
 
-# =============================================================================================
-# Base class for receptor-ligand restraints.
-# =============================================================================================
 
-ABC = abc.ABCMeta('ABC', (object,), {}) # compatible with Python 2 *and* 3
+# ==============================================================================
+# ComposableState class to control the strength of restraints.
+# ==============================================================================
 
+class RestraintState(object):
+    """The state of a restraint.
 
-class ReceptorLigandRestraint(ABC):
-    """
-    Impose a single restraint between ligand and protein to prevent ligand from drifting too far
-    from protein in implicit solvent calculations.
+    A `ComposableState` controlling the strength of a restraint
+    through its `lambda_restraints` property.
 
-    This restraint strength is controlled by a global context parameter called 'lambda_restraints'.
+    Parameters
+    ----------
+    lambda_restraints : float
```

Comment:
Not really without implementing a method in `CompoundThermodynamicState` that does the job. However, all the `lambda_X` parameters will stay in `AlchemicalState` with the exception of `lambda_restraints`. I thought it was cleaner to keep it separated since `alchemy` doesn't implement any restraint-related functionality.